cwms_radar_api/src/main/java/cwms/radar/data/dao/TimeSeriesDaoImpl.java

```diff
@@ -0,0 +1,649 @@
+package cwms.radar.data.dao;
+
+import java.math.BigDecimal;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.TemporalAccessor;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import cwms.radar.data.dto.Catalog;
+import cwms.radar.data.dto.CwmsDTOPaginated;
+import cwms.radar.data.dto.RecentValue;
+import cwms.radar.data.dto.TimeSeries;
+import cwms.radar.data.dto.Tsv;
+import cwms.radar.data.dto.TsvDqu;
+import cwms.radar.data.dto.TsvDquId;
+import cwms.radar.data.dto.TsvId;
+import cwms.radar.data.dto.catalog.CatalogEntry;
+import cwms.radar.data.dto.catalog.TimeseriesCatalogEntry;
+import org.jetbrains.annotations.NotNull;
+import org.jooq.Condition;
+import org.jooq.DSLContext;
+import org.jooq.Field;
+import org.jooq.Record;
+import org.jooq.Record1;
+import org.jooq.Record3;
+import org.jooq.Record5;
+import org.jooq.Result;
+import org.jooq.SQL;
+import org.jooq.SelectConditionStep;
+import org.jooq.SelectHavingStep;
+import org.jooq.SelectJoinStep;
+import org.jooq.SelectSelectStep;
+import org.jooq.conf.ParamType;
+import org.jooq.exception.DataAccessException;
+import org.jooq.impl.DSL;
+
+import usace.cwms.db.dao.ifc.ts.CwmsDbTs;
+import usace.cwms.db.dao.util.services.CwmsDbServiceLookup;
+import usace.cwms.db.jooq.codegen.packages.CWMS_ROUNDING_PACKAGE;
+import usace.cwms.db.jooq.codegen.packages.CWMS_TS_PACKAGE;
+import usace.cwms.db.jooq.codegen.packages.CWMS_UTIL_PACKAGE;
+import usace.cwms.db.jooq.codegen.tables.AV_CWMS_TS_ID2;
+import usace.cwms.db.jooq.codegen.tables.AV_TSV;
+import usace.cwms.db.jooq.codegen.tables.AV_TSV_DQU;
+import usace.cwms.db.jooq.codegen.tables.AV_TS_GRP_ASSGN;
+
+import static org.jooq.impl.DSL.asterisk;
+import static org.jooq.impl.DSL.count;
+import static org.jooq.impl.DSL.field;
+import static org.jooq.impl.DSL.max;
+import static org.jooq.impl.DSL.partitionBy;
+import static usace.cwms.db.jooq.codegen.tables.AV_CWMS_TS_ID2.AV_CWMS_TS_ID2;
+
+public class TimeSeriesDaoImpl extends JooqDao<TimeSeries> implements TimeSeriesDao
+{
+	private static final Logger logger = Logger.getLogger(TimeSeriesDaoImpl.class.getName());
+
+	public static final String STORE_RULE = "DELETE INSERT";
```

Comment:
Let's make an enum for this and make it a query parameter. We've already had performance issues from overusing this rule in the past best not to accidentally repeat it.

Reply:
ok.